pkg/cloud/common/templates.go

```diff
@@ -0,0 +1,79 @@
+package common
+
+import (
+	"bytes"
+	"embed"
+	"fmt"
+	"text/template"
+
+	"k8s.io/klog/v2"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/yaml"
+)
+
+type TemplateValues map[string]interface{}
+
+type TemplateSource struct {
+	ReferenceObject client.Object
+	EmbedFsPath     string
+}
+
+type ObjectTemplate struct {
+	TemplateSource
+	templateContent *template.Template
+}
+
+func (tmpl *ObjectTemplate) render(providerAssets TemplateBasedAssets) (client.Object, error) {
+	buf := &bytes.Buffer{}
+	templateValues, err := providerAssets.GetTemplateValues()
+	if err != nil {
+		return nil, fmt.Errorf("can not get values for template rendering: %s", err)
+	}
+	err = tmpl.templateContent.Execute(buf, templateValues)
+	if err != nil {
+		return nil, fmt.Errorf("can not render template: %s", err)
+	}
+	object := tmpl.ReferenceObject.DeepCopyObject()
+
+	err = yaml.UnmarshalStrict(buf.Bytes(), object)
+	if err != nil {
+		klog.Errorf("Cannot decode data from embedded resource %v: %v", tmpl.EmbedFsPath, err)
+		return nil, err
+	}
+	return object.(client.Object), nil
+}
+
+func RenderTemplates(provider TemplateBasedAssets) ([]client.Object, error) {
```

Comment:
Do we need to use an interface in here? Why not just pass through the two types that we need? Might make testing easier as you then don't need to mock this interface